src/map.mo

```diff
@@ -0,0 +1,305 @@
+
+import Base "base";
+
+module {
+  /// Type of stable data of `StableTrieMap`
+  public type StableData = Base.StableData and {
+    last_empty_node : Nat64;
+    last_empty_leaf : Nat64;
+  };
+  /// Bidirectional enumeration of any keys s in the order they are added.
+  /// For a map from keys to index `Nat` it is implemented as trie in stable memory.
+  /// for a map from index `Nat` to keys the implementation is a consecutive interval of stable memory.
+  ///
+  /// Arguments:
+  /// + `pointer_size` is size of pointer of address space, first bit is reserved for internal use,
+  ///   so max amount of nodes in stable trie is `2 ** (pointer_size * 8 - 1)`. Should be one of 2, 4, 5, 6, 8.
+  /// + `aridity` is amount of children of any non leaf node except in trie. Should be one of 2, 4, 16, 256.
+  /// + `root_aridity` is amount of children of root node.
+  /// + `key_size` and `value_size` are sizes of key and value which should be constant per one instance of `Map`
+  ///
+  /// Example:
+  /// ```motoko
+  /// let e = StableTrie.Map({
+  ///   pointer_size = 2;
+  ///   aridity = 2;
+  ///   root_aridity = null;
+  ///   key_size = 2;
+  ///   value_size = 0;
+  /// });
+  /// ```
+  public class Map(args : Base.Args) {
+    let base : Base.StableTrieBase = Base.StableTrieBase(args);
+    assert args.key_size + args.value_size >= 8;
+
+    var last_empty_node : Nat64 = base.loadMask;
+    var last_empty_leaf : Nat64 = base.loadMask;
+
+    func pushEmptyLeaf(leaves : Base.Region, leaf : Nat64) {
+      Region.storeNat64(leaves.region, base.getLeafOffset(leaf), last_empty_leaf);
+      last_empty_leaf := leaf;
+    };
+
+    func popEmptyLeaf(leaves : Base.Region) : ?Nat64 {
+      if (last_empty_leaf == base.loadMask) return null;
+      let ret = last_empty_leaf;
+
+      last_empty_leaf := Region.loadNat64(leaves.region, base.getLeafOffset(last_empty_leaf));
+      ?ret;
+    };
+
+    func pushEmptyNode(node : Nat64) {
+      base.setChild(node, 0, last_empty_node);
+      last_empty_node := node;
+    };
+
+    func popEmptyNode(nodes : Base.Region) : ?Nat64 {
+      if (last_empty_node == base.loadMask) return null;
+      let ret = last_empty_node;
+      last_empty_node := base.getChild(nodes, last_empty_node, 0);
+      ?ret;
+    };
+
+    /// Add `key` and `value` to enumeration. Returns null if pointer size limit exceeded. Returns `size` if the key in new to the enumeration
+    /// or rewrites value and returns index of key in enumeration otherwise.
+    ///
+    /// Example:
+    /// ```motoko
+    /// let e = StableTrie.Map({
+    ///   pointer_size = 2;
+    ///   aridity = 2;
+    ///   root_aridity = null;
+    ///   key_size = 2;
+    ///   value_size = 1;
+    /// });
+    /// assert(e.put("abc", "a") == ?0);
+    /// assert(e.put("aaa", "b") == ?1);
+    /// assert(e.put("abc", "c") == ?0);
+    /// ```
+    /// Runtime: O(key_size) acesses to stable memeory.
+    public func put(key : Blob, value : Blob) : ?Nat {
```

Comment:
Shall we still return `?Nat` in the Map version? Or just `Bool` and not expose the `Nat`?